pkg/cacheutil/memcached_client.go

```diff
@@ -531,6 +544,35 @@ func (c *memcachedClient) getMultiSingle(ctx context.Context, keys []string) (it
 	return items, err
 }
 
+// sortKeysByServer sorts cache keys within a slice based on which server they are
+// sharded to using a memcache.ServerSelector instance. The keys are ordered so keys
+// on the same server are next to each other. Any errors encountered determining which
+// server a key should be on will result in returning keys unsorted (in the same order
+// they were supplied in). Note that output is not guaranteed to be any particular order
+// *except* that keys sharded to the same server will be together. The order of keys
+// returned may change from call to call.
+func (c *memcachedClient) sortKeysByServer(keys []string) []string {
+	bucketed := make(map[string][]string)
+
+	for _, key := range keys {
+		addr, err := c.selector.PickServer(key)
+		// If we couldn't determine the correct server, return keys in existing order
+		if err != nil {
+			return keys
+		}
+
+		addrString := addr.String()
+		bucketed[addrString] = append(bucketed[addrString], key)
+	}
+
+	var out []string
+	for srv := range bucketed {
+		out = append(out, bucketed[srv]...)
+	}
```

Comment:
Did you consider returning `[][]string` here, to save copying every entry?
(I've seen millions of keys here, when called from Mimir).

Reply:
I hadn't. The batching method needs `[]string` to split up. I'm not sure how I could return `[][]string` here without needing to copy it into a `[]string` eventually.

Reply:
Yes! I thought about that initially and decided to go with the bucketing approach since it's `O(n)` compared to sorting which is `O(n log n)`. I can run some benchmarks if you think we should investigate that?

Reply:
Pushed some benchmarks here: https://github.com/56quarters/thanos/commit/5fc84926a63a7c7d026ac1866b14f1061fe96424

Sorting appears to be slower (slightly slower for a small number of keys, much slower for a large number) and perform more allocations in all but the smallest number of keys.

```
$ go test -cpu=1 -run=XXX -benchmem -bench=BenchmarkMemcachedClient_sortKeysByServer ./pkg/cacheutil/
goos: linux
goarch: amd64
pkg: github.com/thanos-io/thanos/pkg/cacheutil
cpu: Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
BenchmarkMemcachedClient_sortKeysByServer/10_keys                 514545              2259 ns/op             960 B/op         20 allocs/op
BenchmarkMemcachedClient_sortKeysByServer/100_keys                 76731             15666 ns/op            9152 B/op        116 allocs/op
BenchmarkMemcachedClient_sortKeysByServer/1000_keys                 8308            146631 ns/op           81315 B/op       1022 allocs/op
BenchmarkMemcachedClient_sortKeysByServer/10000_keys                 656           2213951 ns/op         1187478 B/op      10036 allocs/op
BenchmarkMemcachedClient_sortKeysByServer2/10_keys                424206              3002 ns/op             680 B/op         28 allocs/op
BenchmarkMemcachedClient_sortKeysByServer2/100_keys                21447             54334 ns/op           12536 B/op        522 allocs/op
BenchmarkMemcachedClient_sortKeysByServer2/1000_keys                2263            538877 ns/op          120551 B/op       5022 allocs/op
BenchmarkMemcachedClient_sortKeysByServer2/10000_keys                205           5624709 ns/op         1199250 B/op      49967 allocs/op
PASS
ok      github.com/thanos-io/thanos/pkg/cacheutil       11.488s
```

Reply:
Another option could be introducing an iterator object here that knows how to iterate over an `[][]string` (or `map[string][]string`) but that seems like a lot to save a few allocations to me.